lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2569,18 +2569,75 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
     nominal->diagnose(diag::property_delegate_reserved_name);
   }
 }
+
 // SWIFT_ENABLE_TENSORFLOW
 /// Returns true if the given type conforms to `Differentiable` in the given
 /// module.
-static bool conformsToDifferentiableInModule(Type type, ModuleDecl *module) {
-  auto &ctx = module->getASTContext();
+static bool conformsToDifferentiable(TypeChecker &TC, Type type,
+                                     DeclContext *DC) {
   auto *differentiableProto =
-      ctx.getProtocol(KnownProtocolKind::Differentiable);
-  return LookUpConformanceInModule(module)(
-      differentiableProto->getDeclaredInterfaceType()->getCanonicalType(),
-      type, differentiableProto).hasValue();
+      TC.Context.getProtocol(KnownProtocolKind::Differentiable);
+  return TC.conformsToProtocol(
+      type, differentiableProto, DC,
+      ConformanceCheckFlags::Used | ConformanceCheckFlags::InExpression)
+          .hasValue();
 };
 
+// SWIFT_ENABLE_TENSORFLOW
+/// Creates a `AutoDiffParameterIndices` for the given function type,
+/// inferring all differentiation parameters.
+/// The differentiation parameters are inferred to be:
+/// - All parameters of the function type that conform to `Differentiable`.
+/// - If the function type's result is a function type, then also all
+///   parameters of the function result type that conform to `Differentiable`.
+static AutoDiffParameterIndices *
+inferParameters(AnyFunctionType *functionType, TypeChecker &TC,
```

Comment:
It would be great to remove/deduplicate `AutoDiffParameterIndicesBuilder::inferParameters` if possible.
I removed one usage from `Builtins.cpp`, but there remains [a usage in `Attr.cpp`](https://github.com/apple/swift/blob/48f7718a3fb53701951c6ecd4c4595e3d3a7aeff/lib/AST/Attr.cpp#L351).
I'm not sure how to deduplicate, so I left `AutoDiffParameterIndicesBuilder::inferParameters` untouched.

cc @rxwei

Reply:
I suspect `inferParameters`'s behavior is still incorrect for that usage. It's printing so even if it behaved incorrectly it might be okay. 

It would be quite a hassle to dedupe, but it'll be really great to make the attribute printing logic not depend on conformance checking. Instead, Sema should provide its own logic for pretty printing wrt parameters.

Reply:
The only usage of pretty printing wrt parameters is in [`static void diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance, ValueDecl *req, const RequirementMatch &match)`](https://github.com/apple/swift/blob/48f7718a3fb53701951c6ecd4c4595e3d3a7aeff/lib/Sema/TypeCheckProtocol.cpp#L2251), in TypeCheckProtocol.cpp.

That function doesn't currently take a `TypeChecker &` instance, so we'd have to edit the function to take a `TypeChecker &` and move pretty printing logic out of `getDifferentiationParametersClauseString` in `Attr.cpp` and create another version taking a `TypeChecker &` in TypeCheckProtocol.cpp.

What are your thoughts? Shall we fix the attribute printing logic in this PR?

Reply:
It might not actually need a TypeChecker anymore. We've been making methods of TypeChecker static, but some places still use a TypeChecker instance while only calling static methods on it.

Reply:
Removed `TypeChecker &` argument from functions in https://github.com/apple/swift/commit/38de6947fc62bc1c734d6f5f6f3359f3b5c93b13.

I'm still working on removing `AutoDiffParameterIndicesBuilder::inferParameters` and refactoring the pretty printing logic. Current progress is: there's a regression regarding [`differentiable_attr_type_checking.swift`](https://github.com/apple/swift/blob/ef437460ccfb561505dcaf4b39292522fa85bf32/test/AutoDiff/differentiable_attr_type_checking.swift#L669):

```swift
protocol DifferentiableInit : Differentiable {
  // expected-note @+2 {{protocol requires initializer 'init(x:y:)' with type '(x: Float, y: Float)'}}
  @differentiable
  init(x: Float, y: Float)

  // expected-note @+2 {{protocol requires initializer 'init(x:y:)' with type '(x: Float, y: Int)'}}
  @differentiable(wrt: x)
  init(x: Float, y: Int)
}
// expected-error @+1 {{does not conform to protocol 'DifferentiableInit'}}
struct DifferentiableInitStruct : DifferentiableInit {
  var x: Float
  var y: Float

  // FIXME(TF-284): Fix unexpected diagnostic.
  // expected-note @+2 {{candidate is missing attribute '@differentiable'}}
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
  init(x: Float, y: Float) {
    self.x = x
    self.y = y
  }

  // FIXME(TF-284): Fix unexpected diagnostic.
  // expected-note @+2 {{candidate is missing attribute '@differentiable'}}
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
  init(x: Float, y: Int) {
    self.x = x
    self.y = Float(y)
  }
}
```

```
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:670:26: error: incorrect message found
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         candidate is missing attribute '@differentiable(wrt: x)'
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:678:26: error: incorrect message found
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         candidate is missing attribute '@differentiable(wrt: (x, y))'
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:694:26: error: incorrect message found
  // expected-note @+1 {{candidate is missing attribute '@differentiable(wrt: x)'}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         candidate is missing attribute '@differentiable'
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:710:6: error: expected note not produced
  // expected-note @+2 {{candidate is missing attribute '@differentiable(wrt: x)}}
~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

There are two unexpected things here:
* [TF-284](https://bugs.swift.org/projects/TF/issues/TF-284): There are double `candidatte is missing attribute` errors on both witness initializers in `struct DifferentiableInitStruct`. This seems to occur because [`WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames)`](https://github.com/apple/swift/blob/ef437460ccfb561505dcaf4b39292522fa85bf32/lib/Sema/TypeCheckProtocol.cpp#L1068) returns two witnesses for each `init(x:y:)` requirement in `protocol DifferentiableInit`, which is unexpected because only one of the witnesses has a matching type. In other words, it seems weird `DifferentiableInitStruct.init(x: Float, y: Int)` is returned as a witness for `DifferentiableInit.init(x: Float, y: Float)`. cc @slavapestov: do you happen to know why this occurs?
* The regression is that attributes now print differently (e.g. once as `@differentiable` and once as `@differentiable(wrt: (x, y))`).

Reply:
Update: I found fixes regarding both issues! Will edit this comment with more information.